### PR TITLE
Add ThreadLocalStorage::get_for_thread().

### LINKED ISSUE #16921: Remove LogStream from the friends of ThreadLocalStorage
This seems like a funny friendship, but it is because `LogStream` has this member function
```
std::stack<std::string> &
LogStream::get_prefixes() const
{
  bool                     exists         = false;
  std::stack<std::string> &local_prefixes = prefixes.get(exists);

  // If this is a new locally stored stack, copy the "blessed" prefixes
  // from the initial thread that created logstream.
  if (!exists)
    {
      auto it = prefixes.data.find(parent_thread);                       // <-------- accesses ThreadLocalStorage::data
      if (it != prefixes.data.end())
        local_prefixes = it->second;
    }

  return local_prefixes;
}
```
to copy the data associated with one thread to the data associated with the current thread. This is probably better done using a different mechanism, such as `ThreadLocalStorage::get(const std::thread::id &)`. 

This would certainly be useful because the current version is not thread-safe: If the thread we're currently looking up (with by default is not the thread we're working on) is changing its prefixes, then the three lines in question are not atomic.

### PR DESCRIPTION
**Pull request by bangerth**
Fixes #16921.

### CODE REVIEW

File: include/deal.II/base/thread_local_storage.h [hunk 1]
```diff
@@ -175,6 +176,23 @@ namespace Threads
...
+     * If the thread with given `id` has an object currently stored, then return
+     * it by value via the `std::optional` object. If the indicated thread does
+     * not have an object stored, return an empty `std::optional`.
+     *
+     * Note that in the successful case, this function returns a *copy* of
+     * the object, unlike get() which returns a reference to it. This is
+     * because when you call get(), you are calling it from the current
+     * thread (i.e., the thread that "owns" the object), and so all accesses
+     * are by definition not concurrent. On the other hand, if you are asking
+     * about the object owned by a different thread, that other thread my
```
Comment (masterleinad) [inline include/deal.II/base/thread_local_storage.h:189]:
```suggestion
     * about the object owned by a different thread, that other thread might
```

File: include/deal.II/base/thread_local_storage.h [hunk 2]
```diff
@@ -256,8 +274,6 @@ namespace Threads
      * An exemplar for creating a new (thread specific) copy.
      */
     std::shared_ptr<const T> exemplar;
-
-    friend class dealii::LogStream;
```
Comment (tamiko) [inline include/deal.II/base/thread_local_storage.h:260]: Unfriending someone on social media is often considered to be mean. Here, I think, it is OK.

**Review by tamiko — APPROVED** [review]
Thanks for cleaning this up.

**Comment by tamiko** [comment]
@bangerth If memory serves me right I think we chose this awkward solution back in the day because there was no `std::thread::id` in C++98.

**Review by masterleinad — APPROVED** [review]

File: include/deal.II/base/thread_local_storage.h (hunk 1 above)
Comment (bangerth) [reply include/deal.II/base/thread_local_storage.h]: I meant to say "may", "might" might even be better. So fixed.

**Review by kronbichler — APPROVED** [review]

**Pull request merged**